### DISCUSSION TITLE
How to handle CS9057 error after adding Microsoft.Extensions.Http.Resilience as a dependency

### DISCUSSION THREAD

**Discussion by g0loob** [dotnet/extensions#5055] [Q&A] [answered]
I have a .NET 6 project and I wanted to add `Microsoft.Extensions.Http.Resilience` as a dependency. After adding, I have following errors:

```shell
CSC : error CS9057: The analyzer assembly '/root/.nuget/packages/microsoft.extensions.telemetry.abstractions/8.3.0/analyzers/dotnet/cs/Microsoft.Gen.Logging.dll' references version '4.4.0.0' of the compiler, which is newer than the currently running version '4.3.0.0'. [/cs9057/Issue/Issue.csproj]

CSC : error CS9057: The analyzer assembly '/root/.nuget/packages/microsoft.extensions.telemetry.abstractions/8.3.0/analyzers/dotnet/cs/Microsoft.Gen.Metrics.dll' references version '4.4.0.0' of the compiler, which is newer than the currently running version '4.3.0.0'. [/cs9057/Issue/Issue.csproj]
```

I found [similar issue](https://github.com/dotnet/extensions/issues/4969), but so far no solution.

---

To reproduce the issue, I prepared a simple `Dockerfile` with required commands:

```dockerfile
FROM mcr.microsoft.com/dotnet/sdk:6.0

WORKDIR cs9057

# create MWE
RUN dotnet new sln \
    && dotnet new classlib -f net6.0 -o Issue \
    && dotnet sln add Issue/Issue.csproj \
    && dotnet add Issue package Microsoft.Extensions.Http.Resilience

RUN dotnet build /WarnAsError
```

Copy into a `Dockefile` file and run `docker build -t cs9057 .`

---

Any ideas how to handle this issue?

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
Can you upgrade to a higher version of .NET SDK?  You'd still be able to target .NET 6.0.

**Reply by g0loob** [reply]
Thanks for the quick response!

Is there a way without upgrading SDK? I need to check if I'm able to upgrade the version. The problem I have appeared in CI pipeline which builds libraries inside a Docker container with .NET 6 SDK and has `/WarnAsError` flag. The pipeline is generalized for all the libraries in the company.

**Reply by KalleOlaviNiemitalo** [reply]
https://github.com/dotnet/sdk/issues/20355 looks related; the Microsoft.Extensions.Telemetry.Abstractions package could be changed to carry analyzers targeting multiple versions of the compiler and choose the compatible one at build time.  However, I don't know whether these analyzers depend on Roslyn APIs that the older compiler doesn't support.

I don't know how necessary the source generators of that package are.  It might be possible to just disable them.  NuGet doesn't directly support disabling source generators or other analyzers (<https://github.com/NuGet/Home/issues/6279>) but perhaps you could do that with a custom MSBuild target, similar to how the Microsoft.Extensions.Telemetry.Abstractions package disables the source generators of the Microsoft.Extensions.Logging.Abstractions package here: <https://github.com/dotnet/extensions/blob/1b7015c80e33f047688c2a266af9f2e2aebb00ab/src/Libraries/Microsoft.Extensions.Telemetry.Abstractions/buildTransitive/net6.0/Microsoft.Extensions.Telemetry.Abstractions.targets#L2-L15>

**Reply by KalleOlaviNiemitalo** [reply]
My question on SDK requirements: <https://github.com/dotnet/extensions/discussions/4736>

**Reply by g0loob** [reply]
Thanks!!! I can confirm that problem is not there if I use .NET 8 SDK.
